File: src/trie/implement_trie/trie.java
class Trie {

    TrieNode root = new TrieNode();

    private class TrieNode {

        Map<Character, TrieNode> children = new HashMap<>();
        boolean isEnd;

        TrieNode() {

        }
    }

    public Trie() {
        
    }
    
    public void insert(String word) {
        TrieNode node = root;
        for (int i = 0; i < word.length(); i++){
            Character ch = word.charAt(i);
            if (!node.children.containsKey(ch)) {
                node.children.put(ch, new TrieNode());
            }
            node = node.children.get(ch);
        }
        node.isEnd = true;
    }
    
    public boolean search(String word) {
        TrieNode node = root;
        for (int i = 0; i < word.length(); i++){
            Character ch = word.charAt(i);
            if (!node.children.containsKey(ch)) {
                return false;
            }
            node = node.children.get(ch);
        }
        return node.isEnd;
    }
    
    public boolean startsWith(String prefix) {
        TrieNode node = root;
        for (int i = 0; i < prefix.length(); i++){
            Character ch = prefix.charAt(i);
            if (!node.children.containsKey(ch)) {
                return false;
            }
            node = node.children.get(ch);
        }
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: src/datastructures/trie.py
class TrieNode:
    def __init__(self) -> None:
        self.children = {}
        self.endWord = False

class Trie:
    def __init__(self) -> None:
        self.root = TrieNode()
    
    def addWord(self, word: str) -> None:
        temp = self.root

        for ch in word:
            if ch in temp.children:
                temp = temp.children[ch]
            else:
                node = TrieNode()
                temp.children[ch] = node
                temp = temp.children[ch]
        temp.endWord = True

    def search(self, word: str) -> None:
        temp = self.root
        for ch in word:
            if ch in temp.children:
                temp = temp.children[ch]
            else:
                return False    
        return temp.endWord
    
    def searchPrefix(self, prefix: str) -> None:
        temp = self.root
        for ch in prefix:
            if ch in temp.children:
                temp = temp.children[ch]
            else:
                return False    
        return True


